[1단계 - 로또 구현] 크로플(이동환) 미션 제출합니다.

perenok opened the pull request:
안녕하세요 게이츠! 
새로운 리뷰어 분은 언제나 환영입니다~~
이번 미션은 갑자기 생각해야 할게 많아져서 시간 안에 한다고 만족스럽지 못한 부분이 많은 거 같아요ㅋㅋㅋ
TDD로 진행하기 위해 먼저 평소와 달리 설계를 고민하게 되었고, 어떤 메소드가 어디랑 연결되는지를 생각하면서 만드니까 나중에 뷰와 컨트롤러를 만들 때 훨씬 편해지더라고요ㅋㅋㅋ 테스트에 대한 믿음을 통해 작동에 대해 불안감도 덜하고요ㅋㅋㅋㅋ
그리고 enum 클래스를 처음 접하고 만들어봐서 어려웠지만 이걸 통해 데이터 관리가 더 쉽고 편하게 작동해서 좋았습니다!
역시 자바에 있는 함수는 많이 이용하면 할수록 좋은거 같아요ㅋㅋㅋㅋ 마지막에 갑자기 에러 나서 급하게 수정한다고 메소드 이름이 정돈되지 않아 아쉽네요ㅠㅠ

궁금한 점
1. .stream().forEach()를 이용하여 반복문을 전개할 때 여기에 조건을 하나 더 추가하는 방법이 있을까요? OutputView에 printRankResult에서 2등일 때 FORM을 다르게 하고 싶었는데 시간 내에 방빕이 생각나지 않아 결국 추가하지 못했습니다ㅠㅠ
2. commit을 할 때 테스트와 도메인 클래스를 한번에 커밋을 해봤는데 혹시 각각 내용에 대해 언급하면서 커밋을 하는게 더 좋은가요?

많은 피드백 부탁드립니다!!!
---------------------------------------------------------------------------------------------------
안녕하세요 게이츠!
피드백 수정 후 다시 리뷰요청이네요ㅎㅎ
리팩토링은 데드라인에 대한 촉박함에 깔끔하지 못했던 LottoController를 다듬고 OutputView의 완성, 그리고 게이츠가 해준 피드백을 반영하는데 중점을 뒀어요ㅎㅎ try-catch 재귀를 위해 코드를 분리하다 보니 분리한 메소드에 대한 작명이 꽤나 고역이었어요ㅋㅋㅋㅋ
이번에 리팩토링하면서 궁금한 점이 있는데요!
첫번째는 OutputView에 forEach를 사용한 부분입니다. 
```
   public static void printRankResult(final LottoGameResult lottoGameResult) {
        lottoGameResult.ranks().entrySet().forEach(entry -> {
            if (entry.getKey() == Rank.SECOND) {
                printRankForm(entry, RANK_BONUS_FORM);
            }
            if (entry.getKey() != Rank.SECOND && entry.getKey() != Rank.NOTHING) {
                printRankForm(entry, RANK_FORM);
            }
        });
    }
```
여기서 forEach는 indent depth로 생각하지 않는 건가요? entry -> {...}를 통해 첫번째 depth로 들어가고 if 분기를 통해 두번째 depth로 들어간다는 느낌이 들어서 이 코드가 1 depth가 맞는지에 대해 고민이 들더라고요. 람다식에서 람다바디를 표현하는 중괄호는 depth라고 생각하지 않아도 되는 건가요?

두번째는 Controller를 따로 만들지 않고 main메소드를 컨트롤러에 적용시켜보는 건 어떠냐는 피드백 그 자체입니다ㅋㅋㅋ 고민을 많이 해봐도 명확하게 답을 내릴 수가 없더라고요ㅋㅋㅋ
저는 습관적으로 Application을 만들어 거기에 main 메소드를 추가하고 첫 시작이 되는 Controller 객체를 불러오는 방식을 주로 사용했어요ㅋㅋㅋ 지금 로또 미션은 main에다 Controller의 역할을 구현해도 괜찮을 것 같은데 컨트롤러가 더 많아지고 프로그램이 복잡해지게 되면 그때 마다 메인메소드를 포함한 클래스에 코드가 추가될 것이고 그렇게 되면 하나의 클래스에 많은 책임이 생기지 않을까요? 혹은 LottoController에 main 메소드를 추가해 구현하더라도 마찬가지로 다른 controller가 추가되면 어디서 시작되는지 한번에 찾기 힘들어지고 하지 않을까요? 질문의 요지를 제가 제대로 잡았는지 모르겠네요ㅠㅠ

perenok committed:
refactor : delete unnecessary files

perenok committed:
docs(README) : include lotto feature list

perenok committed:
docs(todo) : create an undo list for each feature

perenok committed:
docs(todo) : modify list

perenok committed:
feat(Lotto) : implements Lotto

Lotto
 - List<LottoNumber> lottoNumbers를 인스턴스 변수로 가진 클래스
containNumber
 - LottoNumber를 인자로 받아 Lotto가 포함하고 있는지 확인하는 메소드
countOfMatchNumber
 - Lotto를 인자로 받아 일치하는 숫자의 개수를 반환하는 메소드

perenok committed:
feat(LottoNumber) : implements LottoNumber

LottoNumber
 - int number를 인스턴스 변수로 가진 클래스
 - 원시값인 int를 포장한다
validateNumber
 - LottoNumber의 범위(1~45)에 속하는지 검증하는 메소드

perenok committed:
feat(Lottos) : implements Lottos

Lottos
 - 일급 컬렉션
 - List<Lotto> lottos를 유일한 필드로 가짐
add
 - Lotto 객체를 인자로 받아 lottos에 추가하는 메소드
lottos()
 - lottos를 반환하는 메소드

perenok committed:
feat(Rank) : implements Rank

Rank
 - enum 클래스
 - 등수와 상금을 필드로 가진다
rankOf
 - countOfMatch와 bonusNumber를 인자로 받아 해당하는 enum 상수를 반환하는 메소드
 - matchCount 메소드를 이용해 해당하는 enum 상수를 결정
matchCount
 - 인자로 받은 countOfMatch와 필드 countOfMatch가 동일한지 여부 판단하는 메소드

perenok committed:
feat(Money) : implements Money

Money
 - int money를 필드 변수로 가지는 클래스
 - 원시값인 int를 포장
countLotto
 - 구매금액을 통해 구매 가능한 로또의 개수를 반환하는 메소드
validateMoney
 - 입력받은 구매금액에 대한 유효성 검증하는 메소드
 - validateNumber : 입력값이 숫자인지 판별
 - validateEmpty : null값 판별
 - validatePrice : 1000원 이상의 금액인지 판별

perenok committed:
feat(WinningLotto) : implements WinningLotto

WinningLotto
 - Lotto와 LottoNumber를 필드 변수로 가지는 클래스
findRank
 - countOfMatch : lotto와 WinningLotto간 일치하는 개수 파악
 - matchBonusNumber : bonusNumber를 lotto가 포함하고 있는지 파악
 - 위 두개의 변수를 인자로 넘겨 Rank를 반환
validateBonusNumber
 - lotto의 숫자와 bonusNumber가 중복되는지 판별

perenok committed:
docs(todo) : add doing session and move done feature

perenok committed:
feat(LottoGenerator) : implements LottoGenerator

LottoGenerator
 - List<LottoNumber> numbers 필드변수로 가지는 클래스
 - numbers는 1~45까지의 LottoNumber를 인자로 가짐
generate()
 - 생성된 number를 랜덤하게 섞어 6개만 갖고 있는 리스트를 이용해 Lotto 객체 반환

perenok committed:
feat(LottoGame) : implements LottoGame

buyLottos()
 - 구매 가능한 로또 수 만큼 Lotto 객체 생성하여 lottos에 add
 - 이 때 lottoGenerator를 이용해 자동으로 번호 생성

perenok committed:
feat(GameResult) : implements GameResult

GameResult
 - List<Rank> ranks를 필드변수로 가진 클래스
add
 - 인자로 받은 rank를 필드변수에 저장
countOfLotto
 - 총 구입한 로또의 개수를 반환
totalReward
 - 상금의 총합을 반환
calculateProfit
 - 수익률을 반환
countByRank
  - 인자로 받은 Rank와 동일한 등수에 당첨된 로또의 개수를 반환

perenok committed:
feat(LottoNumber) : implements Comparable<LottoNumber> for use compareTo methods

perenok committed:
feat(Rank) : add matchRank method for compare

perenok committed:
refactor(Lottos) : move calculateTotalPrice from LottoGameResult to Lottos

perenok committed:
refactor(LottoGameResult) : rename and change List to Map

Map 자료구조를 이용하여 등수 별 당첨 티켓의 개수를 쉽게 관리 가능

perenok committed:
refactor(InputView) : implement InputView

askMoney
 - 구입 금액을 입력받아 Money로 반환
askLastWinningLotto
 - 지난 주 당첨 번호를 입력받고 WinningLotto의 인자로 반환
 - askBonusNumber를 통해 받은 LottoNumber 또한 인자로 반환
 - 최종 반환은 WinningLotto
askBonusNumber
 - 보너스 볼 번호를 입력받아 LottoNumber로 반환

perenok committed:
refactor(OutputView) : implement OutputView

printMessage
 - 인자로 받은 message 출력
printMessageByFormat
 - format에 따라 message 가공하여 출력
printEachLotto
 - 출력형태에 맞게 각각의 로또를 출력
printOneLotto
 - 로또 하나의 형태를 가공하여 출력
printLottoGameResult
 - 결과 형태에 맞게 출력
printRankResult
 - 각각의 Rank마다 형태에 맞춰 출력

perenok committed:
feat(Lotto) : add validate logic

perenok committed:
refactor(Lottos) : move calculateTotalPrice from Lottos to LottoGameResult

perenok committed:
style(Money) : add error message

perenok committed:
feat(LottoNumber) : add validate methods and change constructor parameter type

validateNumber
 - 숫자인지 검증
validateEmpty
 - null값 검증
validateNumberBound
 - 1~45의 유효한 숫자인지 검증

perenok committed:
refactor(LottoGenerator) : change parameter type and move LOTTO_NUMBER_SIZE const to Lotto

perenok committed:
refactor(LottoGameResult) : change method logic

perenok committed:
refactor(WinningLottoTest) : change Constructor parameter type

perenok committed:
feat(LottoGameController) : implements LottoGameController

LottoGameController
 - 도메인과 뷰를 연결
buyAutoLotto
 - 돈을 입력받아 금액에 맞는 로또들을 반환
makeWinningLotto
 - WinningLotto를 반환
showGameResult
 - 구입한 로또들과 당첨 로또를 비교하여 OutPutView에 인자 반환

perenok committed:
feat(Application) : add main to run program

perenok committed:
refactor(InputView) : modify methods

WinningLotto를 만들 때 로또넘버와 보너스 넘버 받는 것 분리

perenok committed:
feat(LottoGame) : add method

compareWithWinningLotto
 - 구입한 로또와 당첨 로또를 비교하여 lottoGameResult에 결과 저장 후 반환

perenok committed:
refactor(LottoNumber) : add trim to parameter

perenok committed:
refactor(tests) : change parameter

perenok committed:
docs(todo) : complete todo list

perenok committed:
refactor(Money) : switch method order

perenok committed:
refactor(OutputView) : add LOTTO_PURCHASE_FORM

perenok committed:
refactor(Rank) : change enum order

perenok committed:
refactor(WinningLotto) : add Error message

perenok commented:
학습 로그
[TDD] TDD - 5
내용
TDD를 집적 적용하여 개발을 진행함.
원래 절차적으로 구현하던 습관이 있어 도메인과 컨트롤러, 뷰를 번갈아가며 만들었다. 이번에는 먼저 설계에 대해 고민을 하고, 흐름을 파악한 뒤 도메인에 집중해서 테스트 실패, 테스트 성공, 리팩토링의 과정으로 코딩을 하니 모든 도메인을 만든 후 컨트롤러와 뷰를 만드는데 큰 힘을 들이지 않고 조립하는 느낌으로 완성을 할 수 있었다. 비록 처음에 엄청 느리고 망망대해를 헤엄치는 느낌이 들긴 했지만... 조금씩 짝과 맞춰서 진행하다보니 이 방식이 더 좋다는 걸 몸소 느낄 수 있었다.

[JAVA] Java8 stream - 3
내용
for문을 최대한 지양하고  stream을 이용하여 구현을 함.
stream에 익숙치 않아 IntelliJ의 힘을 빌려 stream().을 치고 자동완성을 통해 나온 메소드들을 쭉 읽어보며 공부함
.map()과 .filter()를 이용하여 조건을 지정하고 .count()나 .sum과 같은 메소드를 통해 바로 계산이 가능한 점을 배움

serverwizard commented on src/main/java/lotto/controller/LottoGameController.java:
```diff
@@ -0,0 +1,50 @@
+import lotto.view.InputView;
+import lotto.view.OutputView;
+
+public class LottoGameController {
```
그리고 LottoGameController를 별도로 만드셨던데 main메소드를 컨트롤러 역할로 사용하면은 안될까요?
제가 드리는 코멘트가 절대 정답이라는 건 아니니, 무조건 반영하실 필요는 없어요~

serverwizard commented on src/main/java/lotto/domain/LottoGame.java:
```diff
@@ -0,0 +1,23 @@
+
+import lotto.utils.LottoGenerator;
+
+public class LottoGame {
```
현재 LottoGame은 속성을 하나도 안 가지고 있는데요. 
LottoGame에 필요한 Lottos(List < Lotto >)를 LottoGame의 속성으로 가지는 건 어떨까요?
LottoGame의 메소드 모두 (butLottos( ), compareWithWinningLotto( )) Lottos를 사용하고 있으니까요.

serverwizard commented on src/main/java/lotto/domain/Lottos.java:
```diff
@@ -0,0 +1,39 @@
+import java.util.List;
+import java.util.Objects;
+
+public class Lottos {
```
일급 콜렉션 사용 👍

serverwizard commented on src/main/java/lotto/domain/LottoGame.java:
```diff
@@ -0,0 +1,23 @@
+        LottoGameResult lottoGameResult = new LottoGameResult();
+        for (Lotto lotto : lottos.lottos()) {
+            lottoGameResult.add(winningLotto.findRank(lotto));
+        }
```
구입한 모든 로또와 당첨 로또를 비교하는 작업인 것 같은데, 이를 Lottos에서 할 수는 없을까요?
Lottos가 구입한 모든 로또를 관리하는 객체니까요.

serverwizard commented on src/main/java/lotto/utils/RandomGenerator.java:
```diff
@@ -0,0 +1,6 @@
+package lotto.utils;
+
+public interface RandomGenerator {
```
전략 패턴 사용 👍 

로또 생성 전략이라면 인터페이스명을 LottoGenerator라고 짓고
이를 구현한 다양한 구현체(AutoLottoGenerator, FixedLottoGenerator etc... )를 만드는 방법은 어떤가요?

serverwizard commented on src/main/java/lotto/controller/LottoGameController.java:
```diff
@@ -0,0 +1,50 @@
+        WinningLotto lastWinningLotto =
+                makeWinningLotto(lastWinningLotto(), InputView.askBonusNumber());
+
+        showGameResult(lottos, lastWinningLotto);
```
아래 코드는 로또 게임을 추상화한 LottoGame 객체를 통해서 게임을 진행해보는 방식인데요.

이해를 돕기 위해 제가 빠르게 작성해 본 코드기 때문에, 절대 정답은 아니니 참고만 부탁드릴게요~

다른 것보다 가독성 측면에서 좀 봐주셨으면 좋겠어요~
```
LottoGame lottoGame = new LottoGame()
...
List<Lotto> createdLottos = lottoGame.createLottos(price, new AutoLottoGenerator());
...
LottoGameResult lottoGameResult = lottoGame.calculateYield(winningNumbers, bonusBall);
```

serverwizard reviewed (changes requested):
안녕하세요 크로플! (빌)게이츠 입니다. ^^

로또 구현을 잘 해주셨네요. 👍 
몇 가지 피드백 남겨두었으니 확인 부탁드릴게요

궁금한점 있으시면 편하게 코멘트 또는 DM 남겨주세요

serverwizard commented:
```
.stream().forEach()를 이용하여 반복문을 전개할 때 여기에 조건을 하나 더 추가하는 방법이 있을까요? 
OutputView에 printRankResult에서 2등일 때 FORM을 다르게 하고 싶었는데 시간 내에 방빕이 생각나지 않아 결국 추가하지 못했습니다ㅠㅠ
```
가장 간단한 방법으로는 forEach 문을 돌면서 if문으로 분기 처리하면 될 것 같아요~
```
   entries.forEach(entry -> {
              if( 조건 ) {
                  ....
              }
                 ...
   });
```
참고로 Iterable에 forEach( ) 가 defulate method 로 정의되어 있기 때문에 stream( ) 으로 변환후 forEach( )를 사용하지 않아도 됩니다.



```
commit을 할 때 테스트와 도메인 클래스를 한번에 커밋을 해봤는데 혹시 각각 내용에 대해 언급하면서 커밋을 하는게 더 좋은가요?
```
이건 개인 취향에 따라 달라질 것 같은데요. 개인적으로 저는 production 코드와 테스트 코드를 한 번에 커밋 하는 편이에요!
대신 커밋 로그 작성할 때 `회원 가입 및 TC 작성` 라고 표현을 해줘요.

perenok committed:
refactor(AutoLottoGenerator) : rename Class

perenok committed:
style(Lotto) : 상수와 필드 변수에 공백 추가

perenok committed:
refactor(Lottos) : move compareWithWinningLotto method from LottoGame

perenok committed:
refactor(LottoGenerator) : change RandomGenerator to LottoGenerator

generate 메소드의 시그니쳐를 Object에서 Lotto로 변경

perenok committed:
feat(LottoGame) : add myLottos as field variable and calculateLottoGameResult method

compareWithWinningLotto를 lottos로 옮기고 필드변수로 추가한 myLottos에 메세지를 보내 계산하게 변경
변경점에 맞춰 TC 수정

perenok committed:
refactor(WinningLotto) : add override methods to compare WinningLotto object

perenok committed:
style(Rank) : change reward format to readability

perenok committed:
style(LottoGameResult) : change ranks.get() to countByRank() and add const ADD_ONE

perenok committed:
refactor(LottoGameController) : clean up methods to improve readability

가독성을 높이기 위해 메소드를 정리하고 각각의 예외상황을 재귀로 처리하기 위해 분리

perenok committed:
refactor(InputView) : change method names and signature

LottoGameController의 변경에 맞춰 수정

perenok committed:
refactor(OutputView) : add if branch to printRankResult method

2등의 경우 다른 출력 상수를 사용하기 위해 if 분기 추가 및 코드를 분리하여 메소드 길이를 줄임

perenok committed:
style(LottoNumber) :delete blank

serverwizard commented on src/main/java/lotto/controller/LottoGameController.java:
```diff
@@ -0,0 +1,48 @@
+        buyAutoLotto();
+        Lotto lotto = InputView.askLastWinningLotto();
+        WinningLotto lastWinningLotto = createLastWinningLotto(lotto);
+        showLottoGameResult(lastWinningLotto);
```
제가 기존에 드린 아래 코멘트도 한 번 생각해 보시면 좋을 것 같아요!

아래 코드는 로또 게임을 추상화한 LottoGame 객체를 통해서 게임을 진행해보는 방식인데요.

이해를 돕기 위해 제가 빠르게 작성해 본 코드기 때문에, 절대 정답은 아니니 참고만 부탁드릴게요~

다른 것보다 가독성 측면에서 좀 봐주셨으면 좋겠어요~
```
LottoGame lottoGame = new LottoGame()
...
List<Lotto> createdLottos = lottoGame.createLottos(price, new AutoLottoGenerator());
...
LottoGameResult lottoGameResult = lottoGame.calculateYield(winningNumbers, bonusBall);
```

serverwizard reviewed (approved):
크로플! 구현 잘해주셨네요! 👍

이번 단계는 여기서 merge하도록 할게요.

다음 단계 진행해주시면 될 것 같아요!
수고하셨습니다! 🙏

serverwizard commented:
```
첫번째는 OutputView에 forEach를 사용한 부분입니다.

   public static void printRankResult(final LottoGameResult lottoGameResult) {
        lottoGameResult.ranks().entrySet().forEach(entry -> {
            if (entry.getKey() == Rank.SECOND) {
                printRankForm(entry, RANK_BONUS_FORM);
            }
            if (entry.getKey() != Rank.SECOND && entry.getKey() != Rank.NOTHING) {
                printRankForm(entry, RANK_FORM);
            }
        });
    }
여기서 forEach는 indent depth로 생각하지 않는 건가요? entry -> {...}를 통해 
첫번째 depth로 들어가고 if 분기를 통해 두번째 depth로 들어간다는 느낌이 들어서 이 코드가 1 depth가 맞는지에 대해 고민이 들더라고요. 
람다식에서 람다바디를 표현하는 중괄호는 depth라고 생각하지 않아도 되는 건가요?
```
사실 엄밀히 말하면 2depth가 들어간 것으로 볼 수 있을 것 같아요!
따라서, 메소드 추출을 통해 depth 1개로 줄여볼 수 있을 것 같아요~

```
두번째는 Controller를 따로 만들지 않고 main메소드를 컨트롤러에 적용시켜보는 건 어떠냐는 피드백 그 자체입니다ㅋㅋㅋ 
고민을 많이 해봐도 명확하게 답을 내릴 수가 없더라고요ㅋㅋㅋ
저는 습관적으로 Application을 만들어 거기에 main 메소드를 추가하고 첫 시작이 되는 Controller 객체를 불러오는 방식을 주로 사용했어요ㅋㅋㅋ 
지금 로또 미션은 main에다 Controller의 역할을 구현해도 괜찮을 것 같은데 컨트롤러가 더 많아지고 프로그램이 복잡해지게 되면 
그때 마다 메인메소드를 포함한 클래스에 코드가 추가될 것이고 그렇게 되면 하나의 클래스에 많은 책임이 생기지 않을까요? 
혹은 LottoController에 main 메소드를 추가해 구현하더라도 마찬가지로 다른 controller가 추가되면 어디서 시작되는지 한번에 찾기 
힘들어지고 하지 않을까요? 질문의 요지를 제가 제대로 잡았는지 모르겠네요ㅠㅠ
```
제가 단지 궁금했던 부분은 현재 로또 미션은 컨트롤러가 1개만 필요한 상황이라 main 메소드를 통해서도 충분히 구현할 수 있을 것 같은데 별도 컨트롤러 클래스를 만드셨길래 어떤 의도를 가지고 만드셨는지 궁금했어요! 다른 의미는 없었답니다 ^^

그리고 아래 글은 제가 잘 이해를 못했어요ㅠ
`그때 마다 메인메소드를 포함한 클래스에 코드가 추가될 것이고 그렇게 되면 하나의 클래스에 많은 책임이 생기지 않을까요?`
